src/wasm-ops/binary-op.cpp

```diff
@@ -0,0 +1,100 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT license.
+
+#include "binary-op.h"
+#include "common.h"
+#include "utils/broadcast_utils.h"
+#include "utils/shape_utils.h"
+
+// Core op
+float add_core(const float &a, const float &b) { return a + b; }
+float sub_core(const float &a, const float &b) { return a - b; }
+float mul_core(const float &a, const float &b) { return a * b; }
+float div_core(const float &a, const float &b) { return a / b; }
+float xor_core(const float &a, const float &b) {
+  return static_cast<float>(static_cast<bool>(a) ^ static_cast<bool>(b));
+}
+float or_core(const float &a, const float &b) {
+  return static_cast<float>(static_cast<bool>(a) || static_cast<bool>(b));
+}
+float and_core(const float &a, const float &b) {
+  return static_cast<float>(static_cast<bool>(a) && static_cast<bool>(b));
+}
+float prelu_core(const float &a, const float &b) { return a >= 0 ? a : a * b; }
+
+// Core binary operator implementation
+void binary_f32_imp(const float *input_1, const int32_t &rank_1,
+                    const std::vector<int32_t> &dims_1, const float *input_2,
+                    const int32_t &rank2, const std::vector<int32_t> &dims_2,
+                    float *output, const int32_t &output_length,
+                    const int32_t &output_rank,
+                    const std::vector<int32_t> &output_dims,
+                    float (*core_op)(const float &, const float &)) {
+  const std::vector<int32_t> &strides1 = ShapeUtils::compute_strides(dims_1);
+  const std::vector<int32_t> &strides2 = ShapeUtils::compute_strides(dims_2);
+  const std::vector<int32_t> &output_strides =
+      ShapeUtils::compute_strides(output_dims);
+
+  for (size_t i = 0; i < output_length; ++i) {
+    auto broadcasted_indices = ShapeUtils::offset_to_indices(output_strides, i);
```

Comment:
is it possible to avoid creating `std::vector` instances inside the loop?
the size of those output indices should be determined. can we create vectors outside the loop so that those util functions use (instead of create and return) them